libflux/src/ast/walk/mod.rs

```diff
@@ -231,23 +231,35 @@ impl<'a> Node<'a> {
 // Done is called on that Node to signal that we are done with that Node.
 //
 // If Visit returns None, walk will not recurse on the children.
-pub trait Visitor: Sized {
+//
+// Note: the Rc in visit and done is to allow for multiple ownership of a node, i.e.
+//       a visitor can own a node as well as the walk funciton. This allows
+//       for nodes to persist outside the scope of the walk function and to
+//       be cleaned up once all owners have let go of the reference.
+pub trait Visitor<'a>: Sized {
     // Visit is called for a node.
     // The returned visitor will be used to walk children of the node.
     // If visit returns None, walk will not recurse on the children.
-    fn visit(&self, node: &Node) -> Option<Self>;
+    fn visit(&self, node: Rc<Node<'a>>) -> Option<Self>;
     // Done is called for a node once it has been visited along with all of its children.
-    fn done(&self, _: &Node) {} // default is to do nothing
+    fn done(&self, _: Rc<Node>) {} // default is to do nothing
```

Comment:
missed this the first time, the Node here should have the explicit lifetime `'a` as well to be consistent.